### PR TITLE
Improve Simple Content Access UX

### PR DESCRIPTION
**Pull request by M4rtinK**
Show a proper status message instead of a confusing (even if correct) message about no subscription being attached if install time registration is done in Simple Content Access mode.

This is achieved by parsing the data returned by the register methods of the RHSM DBus, exposing this via a DBus property of the Subscription module, which is then used by the GUI to display the appropriate string.

### CODE REVIEW

**Comment by M4rtinK** [comment]
rhel_90_SCA_UX_fix
This is how the Connect to Red Hat screen looks like with this PR applied after registration in SCA mode.

**Comment by M4rtinK** [comment]
We can't merge this until subscription-manager 1.29.24 has been released an is available n the compose. So setting the *blocked* label.

**Review by poncovka — APPROVED** [review]
Looks good to me.

**Review by jkonecny12 — APPROVED** [review]
I found some nitpicks but other than that it looks good to me.

File: tests/unit_tests/pyanaconda_tests/modules/subscription/test_subscription_tasks.py [hunk 1]
```diff
@@ -1266,6 +1266,92 @@ def test_get_proxy_url(self):
         assert RegisterAndSubscribeTask._get_proxy_url(request_with_proxy_data) == \
             "http://foo_user:foo_password@proxy.example.com:3128"
 
+    def test_registration_data_json_parsing(self):
+        """Test the registration data JSON parsing method of RegisterAndSubscribeTask."""
+        parse_method = RegisterAndSubscribeTask._detect_sca_from_registration_data
+        # the parsing method should be able to survive also getting an empty string
+        # or even None, returning False
+        assert not parse_method("")
```
[the remark this answers is not in the document]
Comment (VladimirSlavik) [reply tests/unit_tests/pyanaconda_tests/modules/subscription/test_subscription_tasks.py:1280]: That doesn't sound right, though. Assert checks a bool value. The function returns a bool. Seems legit, unless you also want to ensure the function returns only bools and nothing else?

**Review by VladimirSlavik — APPROVED** [review]
Mostly looks good to me, +1 to the other reviews.

I am slightly confused about how "simple content access" and "SCA" are mixed pretty freely, I think you coudl have saved a lot of typing by always using SCA, but never mind :)

File: pyanaconda/modules/subscription/runtime.py [hunk 2]
```diff
@@ -777,8 +780,9 @@ class RegisterAndSubscribeTask(Task):
 
     def __init__(self, rhsm_observer, subscription_request, system_purpose_data,
                  registered_callback, registered_to_satellite_callback,
-                 subscription_attached_callback, subscription_data_callback,
-                 satellite_script_callback, config_backup_callback):
+                 simple_content_access_callback, subscription_attached_callback,
+                 subscription_data_callback, satellite_script_callback,
+                 config_backup_callback):
```
[the remark this answers is not in the document]
Comment (M4rtinK) [reply pyanaconda/modules/subscription/runtime.py:785]: Yeah, doing something about this is definitely planned for the upstream version of Satellite support. :)

File: pyanaconda/modules/subscription/runtime.py [hunk 3]
```diff
@@ -841,6 +841,42 @@ def _get_proxy_url(subscription_request):
...
+                # SCA explicitely noted as enabled
+                return True
+            elif content_access_mode == "entitlement":
+                # SCA explicitely not enabled
+                return False
+            else:
+                log.warning("contentAccessMode mode not set to known value:")
+                log.warning(content_access_mode)
+                # unknown mode or missing data -> not SCA
+                return False
```
[the remark this answers is not in the document]
Comment (M4rtinK) [reply pyanaconda/modules/subscription/runtime.py:881]: Past experience with RHSM & the closely related Candlepin and Satellite APIs indicates a reasonable potential for unexpected values coming back from the API & that it's very useful to have debug data available when this happens.

File: pyanaconda/modules/subscription/runtime.py [hunk 4]
```diff
@@ -841,6 +841,42 @@ def _get_proxy_url(subscription_request):
...
+
+        This function checks JSON data describing registration state as returned
+        by the the Register() or RegisterWithActivationKeys() RHSM DBus methods.
+        Based on the value of the "contentAccessMode" key present in a dictionary available
+        under the "owner" top level key.
+
+        :param str registration_data_json: registration data in JSON format
+        :return: True if data inicates SCA enabled, False otherwise
+        """
+        # we can't try to detect SCA mode if we don't have any registration data data
```
[the remark this answers is not in the document]
Comment (M4rtinK) [reply pyanaconda/modules/subscription/runtime.py:862]: Fixed. :)

File: pyanaconda/modules/subscription/subscription.py [hunk 5]
```diff
@@ -666,6 +692,9 @@ def _system_unregistered_callback(self):
         # script, or else it will be run by the target system
         # provisioning task
         self._set_satellite_provisioning_script(None)
+        # also when we are no longer registered then we are are
+        # thus no longer in Simple Content Access mode
+        self.set_simple_content_access_enabled(False)
```
[the remark this answers is not in the document]
Comment (M4rtinK) [reply pyanaconda/modules/subscription/subscription.py:697]: While I'm afraid this is as far as we can reasonably go given the scope and timeline of this fix, there is definitely some room for improvement in the future.

Ideally the SCA mode value would be part of the data describing attached subscriptions *but* that's handled much later in the flow and with a separate task from the registration one, so I didn't find an easy way of putting it there without possibly changing quite a bit of code.

*Still*, there is a a ray of hope in the form of a forthcoming RHSM API improvement that could potentially remove the need to call `AutoAttach()` *and* maybe even `GetPools()`. Then we could just have a single register task, that would do all whats needed and would return all the necessary data in one go. Would could then feed that data into a single DBus struct describing what happened and be done with it, potentially making the code much simpler and smaller. :)

File: pyanaconda/modules/subscription/runtime.py [hunk 6]
```diff
@@ -283,13 +283,14 @@ def run(self):
...
-                                                {},
-                                                locale)
+                registration_data = private_register_proxy.Register(self._organization,
+                                                                    self._username,
+                                                                    self._password,
+                                                                    {},
+                                                                    {},
+                                                                    locale)
                 log.debug("subscription: registered with username and password")
+                return registration_data
```
[the remark this answers is not in the document]
Comment (M4rtinK) [reply pyanaconda/modules/subscription/runtime.py:293]: Oh, right, thanks! Fixed. :)

File: pyanaconda/modules/subscription/runtime.py [hunk 7]
```diff
@@ -283,13 +283,14 @@ def run(self):
                 locale = os.environ.get("LANG", "")
                 private_register_proxy = private_bus.get_proxy(RHSM.service_name,
                                                                RHSM_REGISTER.object_path)
-                private_register_proxy.Register(self._organization,
-                                                self._username,
-                                                self._password,
-                                                {},
-                                                {},
-                                                locale)
+                registration_data = private_register_proxy.Register(self._organization,
```
[the remark this answers is not in the document]
Comment (M4rtinK) [reply pyanaconda/modules/subscription/runtime.py:286]: Indeed, that looks much better. :)

File: pyanaconda/modules/subscription/runtime.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jkonecny12) [reply pyanaconda/modules/subscription/runtime.py:888]: You would still have the log there. It's just one for both branches which seems fine to me? If you want to still have info that `entitlement` is possible than comment in the code should be enough for that.

**Comment by jstodola** [comment]
https://bugzilla.redhat.com/show_bug.cgi?id=2049139 with subscription-manager-1.29.25-1.el9 in ON_QA, so this PR is not blocked by subscription-manager any more.

**Comment by M4rtinK** [comment]
> https://bugzilla.redhat.com/show_bug.cgi?id=2049139 with subscription-manager-1.29.25-1.el9 in ON_QA, so this PR is not blocked by subscription-manager any more.

Indeed, dropping the blocked label.

**Comment by M4rtinK** [comment]
This can be ported to Fedora only once Satellite support has been ported.

**Pull request merged**